File: packages/ts/core/src/testing/fake-provider.ts
import type { ZodType } from "zod";

import type { LLMProvider, StreamChunk } from "../interface.js";
import type { ChatRequest } from "../models/index.js";
import {
  type TokenUsage,
  type ToolInvocationRequest,
  type ToolResult,
  type UnifiedResponse,
  UnifiedResponseSchema,
} from "../models/index.js";

/** Canned configuration for {@link FakeProvider}. All fields have defaults. */
export interface FakeProviderConfig {
  text?: string;
  toolCalls?: ToolInvocationRequest[];
  usage?: TokenUsage;
  chunks?: StreamChunk[];
  /** Object returned (after schema validation) by `generateStructured`. */
  structured?: unknown;
}

const DEFAULT_TEXT = "Hello from the fake provider.";
const DEFAULT_USAGE: TokenUsage = { promptTokens: 1, completionTokens: 1 };

/**
 * In-memory {@link LLMProvider} with deterministic canned responses — no network
 * and no provider SDK. Proves the interface is implementable and serves as the
 * reusable fixture for testing the orchestrator, middleware, SDK, and CLI.
 */
export class FakeProvider implements LLMProvider {
  readonly providerId = "fake";

  private readonly text: string;
  private readonly toolCalls: ToolInvocationRequest[];
  private readonly usage: TokenUsage;
  private readonly chunks: StreamChunk[];
  private readonly structured: unknown;

  constructor(config: FakeProviderConfig = {}) {
    this.text = config.text ?? DEFAULT_TEXT;
    this.toolCalls = config.toolCalls ?? [];
    this.usage = config.usage ?? DEFAULT_USAGE;
    this.structured = config.structured ?? {};
    this.chunks = config.chunks ?? this.defaultChunks();
  }

  /** Split the canned text into two deltas; the last carries usage + finish. */
  private defaultChunks(): StreamChunk[] {
    const mid = Math.ceil(this.text.length / 2);
    return [
      { delta: this.text.slice(0, mid) },
      {
        delta: this.text.slice(mid),
        usage: this.usage,
        finishReason: "stop",
      },
    ];
  }

  async chat(_req: ChatRequest): Promise<UnifiedResponse> {
    return UnifiedResponseSchema.parse({
      text: this.text,
      usage: this.usage,
      toolCalls: [],
      finishReason: "stop",
    });
  }

  async *stream(_req: ChatRequest): AsyncGenerator<StreamChunk, void, unknown> {
    for (const chunk of this.chunks) {
      yield chunk;
    }
  }

  async generateStructured<T>(_req: ChatRequest, schema: ZodType<T>): Promise<T> {
    return schema.parse(this.structured);
  }

  async invokeTool(req: ChatRequest): Promise<UnifiedResponse> {
    const toolCalls =
      this.toolCalls.length > 0
        ? this.toolCalls
        : [
            {
              id: "call_1",
              name: req.tools?.[0]?.name ?? "echo",
              arguments: {},
            },
          ];
    return UnifiedResponseSchema.parse({
      text: "",
      usage: this.usage,
      toolCalls,
      finishReason: "tool_use",
    });
  }

  /**
   * Resume after a halted tool call. Returns a deterministic final answer that
   * echoes the supplied results, mirroring how a real adapter would resume to a
   * `finishReason: "stop"` response once the host hands tool output back.
   */
  async resumeTool(
    _req: ChatRequest,
    results: ToolResult[],
  ): Promise<UnifiedResponse> {
    const summary = results.map((r) => `${r.name}=${r.content}`).join(", ");
    return UnifiedResponseSchema.parse({
      text: summary.length > 0 ? `Resumed with ${summary}.` : this.text,
      usage: this.usage,
      toolCalls: [],
      finishReason: "stop",
    });
  }
}

File: packages/py/llmorch-core/src/llmorch_core/testing/fake_provider.py
"""In-memory FakeProvider — twin of packages/ts/core/src/testing/fake-provider.ts.

Deterministic canned responses, no network, no provider SDK. Proves the
interface is implementable and is the reusable fixture later features test against.
"""

from collections.abc import AsyncIterator
from typing import TypeVar

from pydantic import BaseModel

from ..interface import LLMProvider, StreamChunk
from ..models import (
    ChatRequest,
    TokenUsage,
    ToolInvocationRequest,
    ToolResult,
    UnifiedResponse,
)

T = TypeVar("T", bound=BaseModel)

DEFAULT_TEXT = "Hello from the fake provider."


class FakeProvider(LLMProvider):
    provider_id = "fake"

    def __init__(
        self,
        *,
        text: str = DEFAULT_TEXT,
        tool_calls: list[ToolInvocationRequest] | None = None,
        usage: TokenUsage | None = None,
        chunks: list[StreamChunk] | None = None,
        structured: object | None = None,
    ) -> None:
        self._text = text
        self._tool_calls = tool_calls or []
        self._usage = usage or TokenUsage(prompt_tokens=1, completion_tokens=1)
        self._structured: object = {} if structured is None else structured
        self._chunks = chunks if chunks is not None else self._default_chunks()

    def _default_chunks(self) -> list[StreamChunk]:
        mid = (len(self._text) + 1) // 2
        return [
            StreamChunk(delta=self._text[:mid]),
            StreamChunk(delta=self._text[mid:], usage=self._usage, finish_reason="stop"),
        ]

    async def chat(self, req: ChatRequest) -> UnifiedResponse:
        return UnifiedResponse(
            text=self._text,
            usage=self._usage,
            tool_calls=[],
            finish_reason="stop",
        )

    async def stream(self, req: ChatRequest) -> AsyncIterator[StreamChunk]:
        for chunk in self._chunks:
            yield chunk

    async def generate_structured(self, req: ChatRequest, schema: type[T]) -> T:
        return schema.model_validate(self._structured)

    async def invoke_tool(self, req: ChatRequest) -> UnifiedResponse:
        tool_calls = self._tool_calls or [
            ToolInvocationRequest(
                id="call_1",
                name=req.tools[0].name if req.tools else "echo",
                arguments={},
            )
        ]
        return UnifiedResponse(
            text="",
            usage=self._usage,
            tool_calls=tool_calls,
            finish_reason="tool_use",
        )

    async def resume_tool(self, req: ChatRequest, results: list[ToolResult]) -> UnifiedResponse:
        """Deterministic final answer echoing the supplied results — mirrors how a
        real adapter resumes to a ``finish_reason="stop"`` response once the host
        hands tool output back."""
        summary = ", ".join(f"{r.name}={r.content}" for r in results)
        return UnifiedResponse(
            text=f"Resumed with {summary}." if summary else self._text,
            usage=self._usage,
            tool_calls=[],
            finish_reason="stop",
        )
